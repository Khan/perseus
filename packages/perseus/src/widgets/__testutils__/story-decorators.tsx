import * as React from "react";

export const mobileDecorator = (Story) => (
    <div className="framework-perseus perseus-mobile">
        <Story />
    </div>
);

export const articleDecorator = (Story) => (
    <div className="framework-perseus perseus-article">
        <Story />
    </div>
);

export const mobileArticleDecorator = (Story) => (
    <div className="framework-perseus perseus-mobile perseus-article">
        <Story />
    </div>
);

export const rtlDecorator = (Story) => (
    <div style={{direction: "rtl"}}>
        <Story />
    </div>
);
