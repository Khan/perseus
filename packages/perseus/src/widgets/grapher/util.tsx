/* eslint-disable one-var */
import {point as kpoint} from "@khanacademy/kmath";
import * as React from "react";
import _ from "underscore";

import Graphie from "../../components/graphie";
import {getDependencies} from "../../dependencies";
import Util from "../../util";

import type {Coord} from "../../interactive2/types";
import type {PerseusScore} from "../../types";

// @ts-expect-error - TS2339 - Property 'Plot' does not exist on type 'typeof Graphie'.
const Plot = Graphie.Plot;

export const DEFAULT_BACKGROUND_IMAGE = {
    url: null,
} as const;

// TODO(charlie): These really need to go into a utility file as they're being
// used by both interactive-graph and now grapher.
function canonicalSineCoefficients(coeffs: any) {
    // For a curve of the form f(x) = a * Sin(b * x - c) + d,
    // this function ensures that a, b > 0, and c is its
    // smallest possible positive value.
    let amplitude = coeffs[0];
    let angularFrequency = coeffs[1];
    let phase = coeffs[2];
    const verticalOffset = coeffs[3];

    // Guarantee a > 0
    if (amplitude < 0) {
        amplitude *= -1;
        angularFrequency *= -1;
        phase *= -1;
    }

    const period = 2 * Math.PI;
    // Guarantee b > 0
    if (angularFrequency < 0) {
        angularFrequency *= -1;
        phase *= -1;
        phase += period / 2;
    }

    // Guarantee c is smallest possible positive value
    while (phase > 0) {
        phase -= period;
    }
    while (phase < 0) {
        phase += period;
    }

    return [amplitude, angularFrequency, phase, verticalOffset];
}

function canonicalTangentCoefficients(coeffs: any) {
    // For a curve of the form f(x) = a * Tan(b * x - c) + d,
    // this function ensures that a, b > 0, and c is its
    // smallest possible positive value.
    let amplitude = coeffs[0];
    let angularFrequency = coeffs[1];
    let phase = coeffs[2];
    const verticalOffset = coeffs[3];

    // Guarantee a > 0
    if (amplitude < 0) {
        amplitude *= -1;
        angularFrequency *= -1;
        phase *= -1;
    }

    const period = Math.PI;
    // Guarantee b > 0
    if (angularFrequency < 0) {
        angularFrequency *= -1;
        phase *= -1;
        phase += period / 2;
    }

    // Guarantee c is smallest possible positive value
    while (phase > 0) {
        phase -= period;
    }
    while (phase < 0) {
        phase += period;
    }

    return [amplitude, angularFrequency, phase, verticalOffset];
}

const PlotDefaults = {
    areEqual: function (coeffs1, coeffs2) {
        return Util.deepEq(coeffs1, coeffs2);
    },

    Movable: Plot,

    getPropsForCoeffs: function (coeffs) {
        return {
            // @ts-expect-error - TS2339 - Property 'getFunctionForCoeffs' does not exist on type '{ readonly areEqual: (coeffs1: any, coeffs2: any) => boolean; readonly Movable: any; readonly getPropsForCoeffs: (coeffs: any) => any; }'.
            fn: _.partial(this.getFunctionForCoeffs, coeffs),
        };
    },
} as const;

const Linear = _.extend({}, PlotDefaults, {
    url: "https://ka-perseus-graphie.s3.amazonaws.com/67aaf581e6d9ef9038c10558a1f70ac21c11c9f8.png",

    defaultCoords: [
        [0.25, 0.75],
        [0.75, 0.75],
    ],

    getCoefficients: function (coords) {
        const p1 = coords[0];
        const p2 = coords[1];

        const denom = p2[0] - p1[0];
        const num = p2[1] - p1[1];

        if (denom === 0) {
            return;
        }

        const m = num / denom;
        const b = p2[1] - m * p2[0];
        return [m, b];
    },

    getFunctionForCoeffs: function (coeffs, x) {
        const m = coeffs[0],
            b = coeffs[1];
        return m * x + b;
    },

    getEquationString: function (coords) {
        const coeffs = this.getCoefficients(coords);
        const m = coeffs[0],
            b = coeffs[1];
        return "y = " + m.toFixed(3) + "x + " + b.toFixed(3);
    },
});

const Quadratic = _.extend({}, PlotDefaults, {
    url: "https://ka-perseus-graphie.s3.amazonaws.com/e23d36e6fc29ee37174e92c9daba2a66677128ab.png",

    defaultCoords: [
        [0.5, 0.5],
        [0.75, 0.75],
    ],
    // @ts-expect-error - TS2339 - Property 'Parabola' does not exist on type 'typeof Graphie'.
    Movable: Graphie.Parabola,

    getCoefficients: function (coords) {
        const p1 = coords[0];
        const p2 = coords[1];

        // Parabola with vertex (h, k) has form: y = a * (h - k)^2 + k
        const h = p1[0];
        const k = p1[1];

        // Use these to calculate familiar a, b, c
        const a = (p2[1] - k) / ((p2[0] - h) * (p2[0] - h));
        const b = -2 * h * a;
        const c = a * h * h + k;

        return [a, b, c];
    },

    getFunctionForCoeffs: function (coeffs, x) {
        const a = coeffs[0],
            b = coeffs[1],
            c = coeffs[2];
        return (a * x + b) * x + c;
    },

    getPropsForCoeffs: function (coeffs) {
        return {
            a: coeffs[0],
            b: coeffs[1],
            c: coeffs[2],
        };
    },

    getEquationString: function (coords) {
        const coeffs = this.getCoefficients(coords);
        const a = coeffs[0],
            b = coeffs[1],
            c = coeffs[2];
        return (
            "y = " +
            a.toFixed(3) +
            "x^2 + " +
            b.toFixed(3) +
            "x + " +
            c.toFixed(3)
        );
    },
});

const Sinusoid = _.extend({}, PlotDefaults, {
    url: "https://ka-perseus-graphie.s3.amazonaws.com/3d68e7718498475f53b206c2ab285626baf8857e.png",

    defaultCoords: [
        [0.5, 0.5],
        [0.6, 0.6],
    ],
    // @ts-expect-error - TS2339 - Property 'Sinusoid' does not exist on type 'typeof Graphie'.
    Movable: Graphie.Sinusoid,

    getCoefficients: function (coords) {
        const p1 = coords[0];
        const p2 = coords[1];

        const a = p2[1] - p1[1];
        const b = Math.PI / (2 * (p2[0] - p1[0]));
        const c = p1[0] * b;
        const d = p1[1];

        return [a, b, c, d];
    },

    getFunctionForCoeffs: function (coeffs, x) {
        const a = coeffs[0],
            b = coeffs[1],
            c = coeffs[2],
            d = coeffs[3];
        return a * Math.sin(b * x - c) + d;
    },

    getPropsForCoeffs: function (coeffs) {
        return {
            a: coeffs[0],
            b: coeffs[1],
            c: coeffs[2],
            d: coeffs[3],
        };
    },

    getEquationString: function (coords) {
        const coeffs = this.getCoefficients(coords);
        const a = coeffs[0],
            b = coeffs[1],
            c = coeffs[2],
            d = coeffs[3];
        return (
            "y = " +
            a.toFixed(3) +
            " sin(" +
            b.toFixed(3) +
            "x - " +
            c.toFixed(3) +
            ") + " +
            d.toFixed(3)
        );
    },

    areEqual: function (coeffs1, coeffs2) {
        return Util.deepEq(
            canonicalSineCoefficients(coeffs1),
            canonicalSineCoefficients(coeffs2),
        );
    },
});

const Tangent = _.extend({}, PlotDefaults, {
    url: "https://ka-perseus-graphie.s3.amazonaws.com/7db80d23c35214f98659fe1cf0765811c1bbfbba.png",

    defaultCoords: [
        [0.5, 0.5],
        [0.75, 0.75],
    ],

    getCoefficients: function (coords) {
        const p1 = coords[0];
        const p2 = coords[1];

        const a = p2[1] - p1[1];
        const b = Math.PI / (4 * (p2[0] - p1[0]));
        const c = p1[0] * b;
        const d = p1[1];

        return [a, b, c, d];
    },

    getFunctionForCoeffs: function (coeffs, x) {
        const a = coeffs[0],
            b = coeffs[1],
            c = coeffs[2],
            d = coeffs[3];
        return a * Math.tan(b * x - c) + d;
    },

    getEquationString: function (coords) {
        const coeffs = this.getCoefficients(coords);
        const a = coeffs[0],
            b = coeffs[1],
            c = coeffs[2],
            d = coeffs[3];
        return (
            "y = " +
            a.toFixed(3) +
            " sin(" +
            b.toFixed(3) +
            "x - " +
            c.toFixed(3) +
            ") + " +
            d.toFixed(3)
        );
    },

    areEqual: function (coeffs1, coeffs2) {
        return Util.deepEq(
            canonicalTangentCoefficients(coeffs1),
            canonicalTangentCoefficients(coeffs2),
        );
    },
});

const Exponential = _.extend({}, PlotDefaults, {
    url: "https://ka-perseus-graphie.s3.amazonaws.com/9cbfad55525e3ce755a31a631b074670a5dad611.png",

    defaultCoords: [
        [0.5, 0.55],
        [0.75, 0.75],
    ],

    defaultAsymptote: [
        [0, 0.5],
        [1.0, 0.5],
    ],

    /**
     * Add extra constraints for movement of the points or asymptote (below):
     *   newCoord: [x, y]
     *     The end position of the point or asymptote endpoint
     *   oldCoord: [x, y]
     *     The old position of the point or asymptote endpoint
     *   coords:
     *     An array of coordinates representing the proposed end configuration
     *     of the plot coordinates.
     *   asymptote:
     *     An array of coordinates representing the proposed end configuration
     *     of the asymptote.
     *
     * Return: either a coordinate (to be used as the resulting coordinate of
     * the move) or a boolean, where `true` uses newCoord as the resulting
     * coordinate, and `false` uses oldCoord as the resulting coordinate.
     */
    extraCoordConstraint: function (
        newCoord,
        oldCoord,
        coords,
        asymptote,
        graph,
    ) {
        const y = _.head(asymptote)[1];
        return _.all(coords, (coord) => coord[1] !== y);
    },

    extraAsymptoteConstraint: function (
        newCoord,
        oldCoord,
        coords,
        asymptote,
        graph,
    ) {
        const y = newCoord[1];
        const isValid =
            _.all(coords, (coord) => coord[1] > y) ||
            _.all(coords, (coord) => coord[1] < y);

        if (isValid) {
            return [oldCoord[0], y];
        }
        // Snap the asymptote as close as possible, i.e., if the user moves
        // the mouse really quickly into an invalid region
        const oldY = oldCoord[1];
        const wasBelow = _.all(coords, (coord) => coord[1] > oldY);
        if (wasBelow) {
            const bottomMost = _.min(_.map(coords, (coord) => coord[1]));
            return [oldCoord[0], bottomMost - graph.snapStep[1]];
        }
        const topMost = _.max(_.map(coords, (coord) => coord[1]));
        return [oldCoord[0], topMost + graph.snapStep[1]];
    },

    allowReflectOverAsymptote: true,

    getCoefficients: function (coords, asymptote) {
        const p1 = coords[0];
        const p2 = coords[1];

        const c = _.head(asymptote)[1];
        const b = Math.log((p1[1] - c) / (p2[1] - c)) / (p1[0] - p2[0]);
        const a = (p1[1] - c) / Math.exp(b * p1[0]);
        return [a, b, c];
    },

    getFunctionForCoeffs: function (coeffs, x) {
        const a = coeffs[0],
            b = coeffs[1],
            c = coeffs[2];
        return a * Math.exp(b * x) + c;
    },

    getEquationString: function (coords, asymptote) {
        if (!asymptote) {
            return null;
        }
        const coeffs = this.getCoefficients(coords, asymptote);
        const a = coeffs[0],
            b = coeffs[1],
            c = coeffs[2];
        return (
            "y = " +
            a.toFixed(3) +
            "e^(" +
            b.toFixed(3) +
            "x) + " +
            c.toFixed(3)
        );
    },
});

const Logarithm = _.extend({}, PlotDefaults, {
    url: "https://ka-perseus-graphie.s3.amazonaws.com/f6491e99d34af34d924bfe0231728ad912068dc3.png",

    defaultCoords: [
        [0.55, 0.5],
        [0.75, 0.75],
    ],

    defaultAsymptote: [
        [0.5, 0],
        [0.5, 1.0],
    ],

    extraCoordConstraint: function (
        newCoord,
        oldCoord,
        coords,
        asymptote,
        graph,
    ) {
        const x = _.head(asymptote)[0];
        return (
            _.all(coords, (coord) => coord[0] !== x) &&
            coords[0][1] !== coords[1][1]
        );
    },

    extraAsymptoteConstraint: function (
        newCoord,
        oldCoord,
        coords,
        asymptote,
        graph,
    ) {
        const x = newCoord[0];
        const isValid =
            _.all(coords, (coord) => coord[0] > x) ||
            _.all(coords, (coord) => coord[0] < x);

        if (isValid) {
            return [x, oldCoord[1]];
        }
        // Snap the asymptote as close as possible, i.e., if the user moves
        // the mouse really quickly into an invalid region
        const oldX = oldCoord[0];
        const wasLeft = _.all(coords, (coord) => coord[0] > oldX);
        if (wasLeft) {
            const leftMost = _.min(_.map(coords, (coord) => coord[0]));
            return [leftMost - graph.snapStep[0], oldCoord[1]];
        }
        const rightMost = _.max(_.map(coords, (coord) => coord[0]));
        return [rightMost + graph.snapStep[0], oldCoord[1]];
    },

    allowReflectOverAsymptote: true,

    getCoefficients: function (coords, asymptote) {
        // It's easiest to calculate the logarithm's coefficients by thinking
        // about it as the inverse of the exponential, so we flip x and y and
        // perform some algebra on the coefficients. This also unifies the
        // logic between the two 'models'.
        const flip = (coord: any) => [coord[1], coord[0]];
        const inverseCoeffs = Exponential.getCoefficients(
            _.map(coords, flip),
            _.map(asymptote, flip),
        );
        const c = -inverseCoeffs[2] / inverseCoeffs[0];
        const b = 1 / inverseCoeffs[0];
        const a = 1 / inverseCoeffs[1];
        return [a, b, c];
    },

    getFunctionForCoeffs: function (coeffs, x, asymptote) {
        const a = coeffs[0],
            b = coeffs[1],
            c = coeffs[2];
        return a * Math.log(b * x + c);
    },

    getEquationString: function (coords, asymptote) {
        if (!asymptote) {
            return null;
        }
        const coeffs = this.getCoefficients(coords, asymptote);
        const a = coeffs[0],
            b = coeffs[1],
            c = coeffs[2];
        return (
            "y = ln(" +
            a.toFixed(3) +
            "x + " +
            b.toFixed(3) +
            ") + " +
            c.toFixed(3)
        );
    },
});

const AbsoluteValue = _.extend({}, PlotDefaults, {
    url: "https://ka-perseus-graphie.s3.amazonaws.com/8256a630175a0cb1d11de223d6de0266daf98721.png",

    defaultCoords: [
        [0.5, 0.5],
        [0.75, 0.75],
    ],

    getCoefficients: function (coords) {
        const p1 = coords[0];
        const p2 = coords[1];

        const denom = p2[0] - p1[0];
        const num = p2[1] - p1[1];

        if (denom === 0) {
            return;
        }

        let m = Math.abs(num / denom);
        if (p2[1] < p1[1]) {
            m *= -1;
        }
        const horizontalOffset = p1[0];
        const verticalOffset = p1[1];

        return [m, horizontalOffset, verticalOffset];
    },

    getFunctionForCoeffs: function (coeffs, x) {
        const m = coeffs[0],
            horizontalOffset = coeffs[1],
            verticalOffset = coeffs[2];
        return m * Math.abs(x - horizontalOffset) + verticalOffset;
    },

    getEquationString: function (coords) {
        const coeffs = this.getCoefficients(coords);
        const m = coeffs[0],
            horizontalOffset = coeffs[1],
            verticalOffset = coeffs[2];
        return (
            "y = " +
            m.toFixed(3) +
            "| x - " +
            horizontalOffset.toFixed(3) +
            "| + " +
            verticalOffset.toFixed(3)
        );
    },
});

/* Utility functions for dealing with graphing interfaces. */
const functionTypeMapping = {
    linear: Linear,
    quadratic: Quadratic,
    sinusoid: Sinusoid,
    tangent: Tangent,
    exponential: Exponential,
    logarithm: Logarithm,
    absolute_value: AbsoluteValue,
} as const;

export const allTypes: any = _.keys(functionTypeMapping);

export function functionForType(
    // TODO(jeremy): Actually `$Keys<typeof functionTypeMapping>` but that
    // triggers TypeScript to require all of our Plot types to be fully typed which
    // is a big amount of work/change.
    type: string,
): any {
    return functionTypeMapping[type];
}

export const validate = (state: any, rubric: any): PerseusScore => {
    if (state.type !== rubric.correct.type) {
        return {
            type: "points",
            earned: 0,
            total: 1,
            message: null,
        };
    }

    // We haven't moved the coords
    if (state.coords == null) {
        return {
            type: "invalid",
            message: null,
        };
    }

    // Get new function handler for grading
    const grader = functionForType(state.type);
    const guessCoeffs = grader.getCoefficients(state.coords, state.asymptote);
    const correctCoeffs = grader.getCoefficients(
        rubric.correct.coords,
        rubric.correct.asymptote,
    );

    if (guessCoeffs == null || correctCoeffs == null) {
        return {
            type: "invalid",
            message: null,
        };
    }
    if (grader.areEqual(guessCoeffs, correctCoeffs)) {
        return {
            type: "points",
            earned: 1,
            total: 1,
            message: null,
        };
    }
    return {
        type: "points",
        earned: 0,
        total: 1,
        message: null,
    };
};

export const getEquationString = (props: any): string => {
    const plot = props.plot;
    if (plot.type && plot.coords) {
        const handler = functionForType(plot.type);
        const result = handler.getEquationString(plot.coords, plot.asymptote);
        return result || "";
    }
    return "";
};

export const pointsFromNormalized = (
    coordsList: ReadonlyArray<Coord>,
    range: [Coord, Coord],
    step: [number, number],
    snapStep: [number, number],
): ReadonlyArray<Coord> => {
    const numSteps = function (range: Coord, step: number) {
        return Math.floor((range[1] - range[0]) / step);
    };

    // @ts-expect-error - TS2322 - Type 'number[][]' is not assignable to type 'readonly Coord[]'.
    return coordsList.map((coords) => {
        const unsnappedPoint = coords.map((coord, i) => {
            const currRange = range[i];
            const currStep = step[i];
            const nSteps = numSteps(currRange, currStep);
            const tick = Math.round(coord * nSteps);
            return currRange[0] + currStep * tick;
        });
        // In some graphing widgets, e.g. interactive-graph, you can rely
        // on the Graphie to handle snapping. Here, we need the points
        // returned to already be snapped so that the plot that goes
        // through them is correct.
        return kpoint.roundTo(unsnappedPoint, snapStep);
    });
};

export const maybePointsFromNormalized = (
    coordsList: ReadonlyArray<Coord> | null | undefined,
    range: [Coord, Coord],
    step: [number, number],
    snapStep: [number, number],
): ReadonlyArray<Coord> | null | undefined => {
    if (coordsList) {
        return pointsFromNormalized(coordsList, range, step, snapStep);
    }
    return coordsList;
};

/* Given a plot type, return the appropriate default value for a grapher
 * widget's plot props: type, default coords, default asymptote. */
export const defaultPlotProps = (type: string, graph: any): any => {
    // The coords are null by default, to indicate that the user has not
    // moved them from the default position, and that this widget should
    // therefore be considered empty and ineligible for grading. The user
    // *can* move the coords from the default position and then back if
    // they really want to submit the default coords as their answer, but
    // we currently don't write questions that require this.
    //
    // We *do* write questions in which the asymptote should be left in
    // the default position. For this reason, we fill in the default
    // asymptote rather than leaving it null; if the user moves the coords
    // but not the asymptote, the widget is non-empty and eligible for
    // grading.
    //
    // TODO(mattdr): Consider an updated scoring function that marks the
    // default coords as empty *unless* they're the correct coords. This
    // would remove this default-coords-are-always-wrong constraints on
    // the questions we write, while still maintaining our kind behavior
    // when users forget to update a widget... but we'd also be revealing
    // extra information. It would be valid to always submit the default
    // widget before even reading the question; you can't lose, but you
    // might get a free win.
    const model = functionForType(type);
    const gridStep = [1, 1];
    // @ts-expect-error - TS2345 - Argument of type 'number[]' is not assignable to parameter of type '[number, number]'.
    const snapStep = Util.snapStepFromGridStep(gridStep);
    return {
        type,
        asymptote: maybePointsFromNormalized(
            model.defaultAsymptote,
            graph.range,
            graph.step,
            snapStep,
        ),
        coords: null,
    };
};

/* Given a list of available types, choose which to use. */
export const chooseType: <T>(list: ReadonlyArray<T>) => T | undefined = _.first;

export const getGridAndSnapSteps = (
    options: any,
    boxSize: number,
): {
    gridStep: [number, number];
    snapStep: [number, number];
} => {
    const gridStep =
        options.gridStep ||
        Util.getGridStep(options.range, options.step, boxSize);
    const snapStep = options.snapStep || Util.snapStepFromGridStep(gridStep);
    return {
        gridStep: gridStep,
        snapStep: snapStep,
    };
};

const defaultGraph: {
    labels: ReadonlyArray<string>;
    range: [Coord, Coord];
    step: [number, number];
    backgroundImage: any;
    markings: string;
    rulerLabel: string;
    rulerTicks: number;
    valid: boolean;
    showTooltips: boolean;
} = {
    labels: ["x", "y"],
    range: [
        [-10, 10],
        [-10, 10],
    ],
    step: [1, 1],
    backgroundImage: DEFAULT_BACKGROUND_IMAGE,
    markings: "graph",
    rulerLabel: "",
    rulerTicks: 10,
    valid: true,
    showTooltips: false,
};
const defaultPlot = defaultPlotProps("linear", defaultGraph);

export const DEFAULT_GRAPHER_PROPS: any = {
    graph: defaultGraph,
    plot: defaultPlot,
    availableTypes: [defaultPlot.type],
};

export const typeToButton = (type: string): any => {
    const capitalized = type.charAt(0).toUpperCase() + type.substring(1);
    const staticUrl = getDependencies().staticUrl;

    return {
        value: type,
        title: capitalized,
        content: (
            <img src={staticUrl(functionForType(type).url)} alt={capitalized} />
        ),
    };
};
