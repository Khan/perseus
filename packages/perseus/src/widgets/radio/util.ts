import * as i18n from "@khanacademy/wonder-blocks-i18n";

/**
 * Given a choice's position in the radio widget, return the corresponding
 * "choice letter". (For example, in English, `getChoiceLetter(0)` is "A",
 * `getChoiceLetter(1)` is "B", etc.)
 */
export function getChoiceLetter(pos: number): string {
    /* I18N: This is a list of single-character labels that will appear in
     * front of multiple-choice options. For instance, a multiple-choice
     * question with three options would display
     *  (A) first option
     *  (B) second option
     *  (C) third option
     * There must be spaces between each of the different characters. The
     * characters will show up next to options in the order that they are
     * listed here. Most multiple choice questions have 5 or fewer options.
     */
    const lettersString = i18n._(
        "A B C D E F G H I J K L M N O P Q R S T U V W X Y Z",
    );

    const letters = lettersString.split(" ");

    if (pos < letters.length) {
        // If the position we need is listed in the localized string, use that.
        return letters[pos];
    }
    // If we're out of letters, give up and return a space.
    return " ";
}
