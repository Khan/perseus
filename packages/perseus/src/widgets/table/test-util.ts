import type {PerseusRenderer} from "@khanacademy/perseus-core";

export function generateTableRenderer(
    extend: Partial<PerseusRenderer> = {},
): PerseusRenderer {
    const base: PerseusRenderer = {
        content: "The answer is 42\n[[â˜ƒ table 1]]",
        widgets: {
            "table 1": {
                type: "table",
                options: {
                    headers: ["Column 1", "Column 2"],
                    rows: 2,
                    columns: 2,
                    answers: [
                        ["42", "42"],
                        ["42", "42"],
                    ],
                },
            },
        },
        images: {},
    };

    return {...base, ...extend};
}
