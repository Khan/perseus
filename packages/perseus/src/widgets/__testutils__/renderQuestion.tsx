import {RenderStateRoot} from "@khanacademy/wonder-blocks-core";
import {render} from "@testing-library/react";
import * as React from "react";

import {
    testDependenciesV2,
    testDependencies,
    // eslint-disable-next-line import/no-relative-packages
} from "../../../../../testing/test-dependencies";
import {
    DependenciesContext,
    useDependencies,
    setDependencies,
} from "../../dependencies";
import * as Perseus from "../../index";
import {mockStrings} from "../../strings";
import {registerAllWidgetsForTesting} from "../../util/register-all-widgets-for-testing";

import type {APIOptions} from "../../types";
import type {PerseusRenderer} from "@khanacademy/perseus-core";
import type {PropsFor} from "@khanacademy/wonder-blocks-core";

type RenderResult = ReturnType<typeof render>;

type ExtraProps = Omit<PropsFor<typeof Perseus.Renderer>, "strings">;

export const renderQuestion = (
    question: PerseusRenderer,
    apiOptions: APIOptions = Object.freeze({}),
    extraProps?: ExtraProps,
): {
    container: HTMLElement;
    renderer: Perseus.Renderer;
    rerender: (question: PerseusRenderer, extraProps?: ExtraProps) => void;
    unmount: RenderResult["unmount"];
} => {
    setDependencies(testDependencies);
    registerAllWidgetsForTesting();

    let renderer: Perseus.Renderer | null = null;
    const {container, rerender, unmount} = render(
        <RenderStateRoot>
            <DependenciesContext.Provider value={testDependenciesV2}>
                <Renderer
                    ref={(node) => (renderer = node)}
                    question={question as any}
                    apiOptions={apiOptions}
                    extraProps={{
                        ...extraProps,
                        strings: mockStrings,
                    }}
                />
            </DependenciesContext.Provider>
        </RenderStateRoot>,
    );
    // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
    if (!renderer) {
        throw new Error(`Failed to render!`);
    }
    const renderAgain = (
        question: PerseusRenderer,
        extraProps?: ExtraProps,
    ) => {
        rerender(
            <RenderStateRoot>
                <DependenciesContext.Provider value={testDependenciesV2}>
                    <Renderer
                        ref={(node) => (renderer = node)}
                        question={question}
                        apiOptions={apiOptions}
                        extraProps={{
                            ...extraProps,
                            strings: mockStrings,
                        }}
                    />
                </DependenciesContext.Provider>
            </RenderStateRoot>,
        );
        if (!renderer) {
            throw new Error(`Failed to rerender!`);
        }
    };

    return {container, renderer, rerender: renderAgain, unmount};
};

const Renderer = React.forwardRef<
    Perseus.Renderer,
    {
        question: PerseusRenderer;
        apiOptions: APIOptions;
        extraProps?: PropsFor<typeof Perseus.Renderer>;
    }
>((props, ref) => {
    const dependencies = useDependencies();
    return (
        <Perseus.Renderer
            ref={ref}
            content={props.question.content}
            images={props.question.images}
            widgets={props.question.widgets}
            problemNum={0}
            apiOptions={props.apiOptions}
            strings={mockStrings}
            {...props.extraProps}
            {...dependencies}
        />
    );
});
