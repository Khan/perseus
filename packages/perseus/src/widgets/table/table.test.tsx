import {splitPerseusItem, type PerseusItem} from "@khanacademy/perseus-core";
import {scorePerseusItem} from "@khanacademy/perseus-score";
import {screen} from "@testing-library/react";
import {userEvent as userEventLib} from "@testing-library/user-event";

import {generateTestPerseusItem} from "../../util/test-utils";
import {renderQuestion} from "../__testutils__/renderQuestion";

import {generateTableRenderer} from "./test-util";

import type {UserEvent} from "@testing-library/user-event";

function getFullItem(): PerseusItem {
    return generateTestPerseusItem({question: generateTableRenderer()});
}

function getSplitItem(): PerseusItem {
    const item = getFullItem();
    item.question = splitPerseusItem(item.question);
    return item;
}

describe("table", () => {
    let userEvent: UserEvent;
    beforeEach(() => {
        userEvent = userEventLib.setup({
            advanceTimers: jest.advanceTimersByTime,
        });
    });

    describe.each([
        {optionsMode: "answerful", renderItem: getFullItem()},
        {optionsMode: "answerless", renderItem: getSplitItem()},
    ])("answerful vs answerless", ({optionsMode, renderItem}) => {
        it(`${optionsMode}: renders`, () => {
            renderQuestion(renderItem.question);

            expect(screen.getByText("Column 1")).toBeInTheDocument();
            expect(screen.getByText("Column 2")).toBeInTheDocument();
        });

        it(`${optionsMode}: snapshots`, () => {
            const {container} = renderQuestion(renderItem.question);

            expect(container).toMatchSnapshot();
        });

        it(`${optionsMode}: can be answered`, async () => {
            const {renderer} = renderQuestion(renderItem.question);

            const inputs = screen.getAllByRole("textbox");
            for (let i = 0; i < 4; i++) {
                await userEvent.type(inputs[i], "8675309");
            }

            expect(renderer.getUserInputMap()).toEqual({
                "table 1": [
                    ["8675309", "8675309"],
                    ["8675309", "8675309"],
                ],
            });
        });

        it(`${optionsMode}: can be scored`, async () => {
            const {renderer} = renderQuestion(renderItem.question);

            const inputs = screen.getAllByRole("textbox");
            for (let i = 0; i < 4; i++) {
                await userEvent.type(inputs[i], "42");
            }

            const userInput = renderer.getUserInputMap();
            const answerful = generateTableRenderer();
            const score = scorePerseusItem(answerful, userInput, "en");

            expect(score).toHaveBeenAnsweredCorrectly();
        });

        it(`${optionsMode}: returns user input in correct order`, async () => {
            const {renderer} = renderQuestion(renderItem.question);

            const inputs = screen.getAllByRole("textbox");
            for (let i = 0; i < 4; i++) {
                await userEvent.type(inputs[i], `${i}`);
            }

            const userInput = renderer.getUserInputMap();
            expect(userInput).toEqual({
                "table 1": [
                    ["0", "1"],
                    ["2", "3"],
                ],
            });
        });
    });
});
