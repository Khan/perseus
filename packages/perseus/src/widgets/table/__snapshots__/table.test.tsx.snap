// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`table answerful vs answerless answerful: snapshots 1`] = `
<div>
  <div
    class="perseus-renderer perseus-renderer-responsive"
  >
    <div
      class="paragraph"
      data-perseus-paragraph-index="0"
    >
      <div
        class="paragraph"
      >
        The answer is 42

        <div
          class="perseus-widget-container widget-nohighlight widget-block"
        >
          <table
            class="perseus-widget-table-of-values non-markdown"
          >
            <thead>
              <tr>
                <th>
                  <div
                    class="perseus-renderer perseus-renderer-responsive"
                  >
                    <div
                      class="paragraph"
                      data-perseus-paragraph-index="0"
                    >
                      <div
                        class="paragraph"
                      >
                        Column 1
                      </div>
                    </div>
                  </div>
                </th>
                <th>
                  <div
                    class="perseus-renderer perseus-renderer-responsive"
                  >
                    <div
                      class="paragraph"
                      data-perseus-paragraph-index="0"
                    >
                      <div
                        class="paragraph"
                      >
                        Column 2
                      </div>
                    </div>
                  </div>
                </th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>
                  <input
                    type="text"
                    value=""
                  />
                </td>
                <td>
                  <input
                    type="text"
                    value=""
                  />
                </td>
              </tr>
              <tr>
                <td>
                  <input
                    type="text"
                    value=""
                  />
                </td>
                <td>
                  <input
                    type="text"
                    value=""
                  />
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`table answerful vs answerless answerless: snapshots 1`] = `
<div>
  <div
    class="perseus-renderer perseus-renderer-responsive"
  >
    <div
      class="paragraph"
      data-perseus-paragraph-index="0"
    >
      <div
        class="paragraph"
      >
        The answer is 42

        <div
          class="perseus-widget-container widget-nohighlight widget-block"
        >
          <table
            class="perseus-widget-table-of-values non-markdown"
          >
            <thead>
              <tr>
                <th>
                  <div
                    class="perseus-renderer perseus-renderer-responsive"
                  >
                    <div
                      class="paragraph"
                      data-perseus-paragraph-index="0"
                    >
                      <div
                        class="paragraph"
                      >
                        Column 1
                      </div>
                    </div>
                  </div>
                </th>
                <th>
                  <div
                    class="perseus-renderer perseus-renderer-responsive"
                  >
                    <div
                      class="paragraph"
                      data-perseus-paragraph-index="0"
                    >
                      <div
                        class="paragraph"
                      >
                        Column 2
                      </div>
                    </div>
                  </div>
                </th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>
                  <input
                    type="text"
                    value=""
                  />
                </td>
                <td>
                  <input
                    type="text"
                    value=""
                  />
                </td>
              </tr>
              <tr>
                <td>
                  <input
                    type="text"
                    value=""
                  />
                </td>
                <td>
                  <input
                    type="text"
                    value=""
                  />
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</div>
`;
